CAP-0023 implementation

sisuresh opened the pull request:
# Description

Implements https://github.com/stellar/stellar-protocol/blob/master/core/cap-0023.md


# Checklist
- [ ] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [ ] Rebased on top of master (no merge commits)
- [ ] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [ ] Compiles
- [ ] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

jonjove commented on src/invariant/test/BucketListIsConsistentWithDatabaseTests.cpp:
```diff
@@ -416,6 +416,21 @@ class ApplyBucketsWorkModifyEntry : public ApplyBucketsWork
     }
 
+    void
+    modifyClaimableBalanceEntry(LedgerEntry& entry)
```
"BucketListIsConsistentWithDatabase modified entries" doesn't work. I think you got this backwards: it should say `entry.data.claimableBalance().balanceID = cb.balanceID` on line 426, and 427-431 should be deleted (the point is to modify the entry but keep the same key).

sisuresh replied:
Fixed. Didn't realize a lot of these were acceptance tests.

sisuresh commented on src/transactions/ClaimClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,185 @@
+validatePredicate(ClaimPredicate const& pred, TimePoint closeTime)
+{
+    switch (pred.type())
+    {
```
This was a mistake. I'll make sure to add `CLAIM_PREDICATE_UNCONDITIONAL` to all the switch statements and test cases.

sisuresh commented on src/transactions/ClaimClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,185 @@
+    case CLAIM_PREDICATE_AND:
+    {
+        auto const& andPredicates = pred.andPredicates();
+        assert(andPredicates.size() == 2);
```
Good point. Changed from `[]` to `at()`.

sisuresh commented on src/transactions/ClaimClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,185 @@
+        return pred.absAfter() < closeTime;
+    case CLAIM_PREDICATE_BEFORE_RELATIVE_TIME:
+    case CLAIM_PREDICATE_AFTER_RELATIVE_TIME:
+        // TODO:shouldn't be able to get here... is aborting fine?
```
Added a throw here for now. Using a different `ClaimPredicate` sounds interesting though. I'm fine with either solution.

MonsieurNicolas replied:
btw in other places in the code we use `abort()` in this situation - I agree thought that in general it's better practice to throw instead instead of crashing (maybe worth creating an issue for this to clean up)

-> default case should throw

sisuresh commented on src/transactions/ClaimClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,185 @@
+    auto const& claimableBalance = ltxEntry.current().data.claimableBalance();
+
+    size_t i = 0;
+    for (; i < claimableBalance.claimants.size(); ++i)
```
done

sisuresh commented on src/transactions/ClaimClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,185 @@
+
+    // Try to give reserve back to createdBy account. If we can't, try to give
+    // it back to sourceAccount
+    auto createdByAccount =
```
I followed your suggestion, but used a `loadAccountAndAddBalance` method that takes the account to load.

sisuresh commented on src/transactions/ClaimClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,185 @@
+    auto amount = claimableBalance.amount;
+    if (asset.type() == ASSET_TYPE_NATIVE)
+    {
+        if (claimableBalance.createdBy == getSourceID())
```
done

sisuresh commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,270 @@
+    }
+
+    switch (pred.type())
+    {
```
fixed

sisuresh commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,270 @@
+    case CLAIM_PREDICATE_AND:
+    {
+        auto& andPredicates = pred.andPredicates();
+        assert(andPredicates.size() == 2);
```
done

sisuresh commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,270 @@
+    {
+        auto relBefore = pred.relBefore();
+        pred.type(CLAIM_PREDICATE_BEFORE_ABSOLUTE_TIME);
+        pred.absBefore() = closeTime > INT64_MAX - relBefore
```
Made a `relativeToAbsolute` method.

sisuresh commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,270 @@
+updatePredicatesForApply(ClaimPredicate& pred, TimePoint closeTime)
+{
+    switch (pred.type())
+    {
```
fixed

sisuresh commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,270 @@
+        return false;
+    }
+
+    assert(addBalance(header, sourceAccount, -entryCost));
```
Wow TIL. Fixed.

sisuresh commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,270 @@
+            return false;
+        }
+
+        assert(addBalance(header, sourceAccount, -amount));
```
fixed

sisuresh commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,270 @@
+    for (auto const& claimant : claimants)
+    {
+        auto const& dest = claimant.v0().destination;
+        if (dests.count(dest) == 0)
```
done

MonsieurNicolas commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,270 @@
+    case CLAIM_PREDICATE_AND:
+    {
+        auto const& andPredicates = pred.andPredicates();
+        if (andPredicates.size() < 2)
```
for clarity `!= 2` might be better here

sisuresh replied:
I agree. I'll also update the cap.

MonsieurNicolas commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,270 @@
+    case CLAIM_PREDICATE_OR:
+    {
+        auto const& orPredicates = pred.orPredicates();
+        if (orPredicates.size() < 2)
```
for clarity `!= 2` might be better here

sisuresh replied:
I agree

MonsieurNicolas commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,274 @@
+
+    OperationID operationID;
+    operationID.type(ENVELOPE_TYPE_OP_ID);
+    operationID.id().sourceAccount = toMuxedAccount(getSourceID());
```
I think this is wrong: you're supposed to use the sourceID from the transaction not from this operation. 

The CAP is actually confusing as it's using "sourceAccount".

Reason you need this to be the tx source account is that otherwise the operationID will not be unique (so we use `txSrcAccount | seqNum` to ensure that it's the case.

Without this change, the call to `create` below will throw with `std::runtime_error("Key already exists")`

MonsieurNicolas replied:
you should add test cases for this

jonjove replied:
@sisuresh Can you also update the CAP to increase clarity about this?

sisuresh replied:
Good catch. Added a test case under section `validate Tx account is used in hash`. I'll also update the cap.

MonsieurNicolas commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,274 @@
+CreateClaimableBalanceOpFrame::insertLedgerKeysToPrefetch(
+    std::unordered_set<LedgerKey>& keys) const
+{
+    // Prefetch issuer for non-native assets
```
comment is wrong, you're prefetching the trust line

sisuresh replied:
fixed

MonsieurNicolas commented on src/transactions/ClaimClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,164 @@
+{
+    return true;
+}
+
```
shouldn't there be an override for `insertLedgerKeysToPrefetch` that loads the claimable entry?

sisuresh replied:
added the override

MonsieurNicolas commented on src/transactions/OperationFrame.cpp:
```diff
@@ -47,7 +49,7 @@ getNeededThreshold(LedgerTxnEntry const& account, ThresholdLevel const level)
 shared_ptr<OperationFrame>
 OperationFrame::makeHelper(Operation const& op, OperationResult& res,
-                           TransactionFrame& tx)
+                           TransactionFrame& tx, size_t index)
```
per my other comment elsewhere, `size_t` is overkill for this as we later have to downcast it. Better to just downcast it once (here). Not sure why it was `size_t` before in `TransactionFrame::makeOperation`.

sisuresh replied:
fixed

MonsieurNicolas commented on src/invariant/BucketListIsConsistentWithDatabase.cpp:
```diff
@@ -134,6 +135,9 @@ BucketListIsConsistentWithDatabase::checkOnBucketApply(
                     ++nData;
                     break;
+                case CLAIMABLE_BALANCE:
+                    ++nClaimableBalance;
```
you forgot to actually check against the database using `countObjects` (line 185)

sisuresh replied:
Ah good catch. Fixed.

MonsieurNicolas commented on src/main/Config.cpp:
```diff
@@ -25,7 +25,7 @@
 namespace stellar
 {
-const uint32 Config::CURRENT_LEDGER_PROTOCOL_VERSION = 13;
+const uint32 Config::CURRENT_LEDGER_PROTOCOL_VERSION = 14;
```
this commit should probably be first as you already reference protocol 14 in previous commits

sisuresh replied:
Moved to the first commit

MonsieurNicolas commented on src/invariant/LedgerEntryIsValid.cpp:
```diff
@@ -226,4 +228,28 @@ LedgerEntryIsValid::checkIsValid(DataEntry const& de, uint32 version) const
+LedgerEntryIsValid::checkIsValid(ClaimableBalanceEntry const& cbe,
+                                 uint32 version) const
+{
+    if (cbe.claimants.empty())
```
Shouldn't this check the rest?

The code here would perform additional checks that we don't use invalid predicate type (like relative timebounds).

sisuresh replied:
That's a good point. Added validation on the predicates.

sisuresh commented on src/transactions/ClaimClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,164 @@
+    if (!loadAccountAndAddBalance(ltx, header, claimableBalance.createdBy,
+                                  claimableBalance.reserve))
+    {
+        if (!loadAccountAndAddBalance(ltx, header, getSourceID(),
```
Yeah the test case is under the section `createdBy == claimant`.

MonsieurNicolas replied:
Ah yes this is correct, the entry goes out of scope and everything is fine (forgot about this).

Note that the test doesn't actually check this particular condition
 the section `createdBy == claimant` only contains a successful scenario and doesn't test the situation when the claimant cannot receive the reserve (will fail with `CLAIM_CLAIMABLE_BALANCE_LINE_FULL`).

MonsieurNicolas replied:
strike that, there are two sections that deal with the `==` case, so we're good

sisuresh commented on src/xdr/Stellar-ledger-entries.x:
```diff
@@ -260,6 +261,91 @@ struct DataEntry
+case CLAIM_PREDICATE_AFTER_ABSOLUTE_TIME:
+    int64 absAfter;
+case CLAIM_PREDICATE_BEFORE_RELATIVE_TIME:
+    int64 relBefore;    // Seconds since closeTime of the ledger in which the
```
done

sisuresh commented on src/test/TestAccount.cpp:
```diff
@@ -229,6 +230,54 @@ TestAccount::bumpSequence(SequenceNumber to)
+    REQUIRE(claimableBalance.amount == amount);
+    REQUIRE(claimableBalance.balanceID == balanceID);
+    REQUIRE(claimableBalance.createdBy == getPublicKey());
+    REQUIRE(claimableBalance.reserve ==
```
Should be fixed.

sisuresh commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,274 @@
+            return false;
+        }
+
+        auto ok = addBalance(header, sourceAccount, -amount);
```
fixed

sisuresh commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,274 @@
+    operationID.type(ENVELOPE_TYPE_OP_ID);
+    operationID.id().sourceAccount = toMuxedAccount(getSourceID());
+    operationID.id().seqNum = mParentTx.getSeqNum();
+    operationID.id().opNum = mOpIndex;
```
fixed

sisuresh commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,274 @@
+int64_t
+relativeToAbsolute(TimePoint closeTime, int64_t relative)
+{
+    return closeTime > (INT64_MAX - relative) ? INT64_MAX
```
Should be fixed.

sisuresh commented on src/transactions/ClaimClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    case CLAIM_PREDICATE_BEFORE_ABSOLUTE_TIME:
+        return pred.absBefore() > closeTime;
```
fixed

sisuresh commented on src/transactions/test/ClaimableBalanceTests.cpp:
```diff
@@ -0,0 +1,993 @@
+    auto claimAccBalanceAfterCreate = getAccAssetBalance(claimAcc);
+
+    auto reserve = claimants.size() * lm.getLastReserve();
+    REQUIRE(createAccNativeBeforeCreate - (isNative ? amount : 0) - reserve -
```
These should be fixed

sisuresh commented on src/ledger/LedgerTxnClaimableBalanceSQL.cpp:
```diff
@@ -0,0 +1,445 @@
+    std::string coll = mDatabase.getSimpleCollationClause();
+
+    mDatabase.getSession() << "DROP TABLE IF EXISTS claimablebalance;";
+    mDatabase.getSession() << "CREATE TABLE claimablebalance ("
```
done

sisuresh commented on src/ledger/LedgerTxnClaimableBalanceSQL.cpp:
```diff
@@ -0,0 +1,445 @@
+// Copyright 2020 Stellar Development Foundation and contributors. Licensed
```
fixed

sisuresh commented on src/transactions/ClaimClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,185 @@
+
+    case CLAIM_PREDICATE_BEFORE_ABSOLUTE_TIME:
+        return pred.absBefore() > closeTime;
+    case CLAIM_PREDICATE_AFTER_ABSOLUTE_TIME:
```
fixed

sisuresh commented on src/transactions/ClaimClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,164 @@
+    auto amount = claimableBalance.amount;
+    if (asset.type() == ASSET_TYPE_NATIVE)
+    {
+        if (!loadAccountAndAddBalance(ltx, header, getSourceID(), amount))
```
This doesn't throw. The test case under the section `createdBy == claimant` tests this..

sisuresh commented on src/transactions/ClaimClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,164 @@
+        }
+    }
+
+    // remove claimable balance entry
```
Changed to `claimableBalanceLtxEntry`.

MonsieurNicolas commented on src/ledger/LedgerTxnClaimableBalanceSQL.cpp:
```diff
@@ -0,0 +1,445 @@
+    mDatabase.getSession() << "CREATE TABLE claimablebalance ("
+                           << "balanceid    VARCHAR(56) " << coll
+                           << " PRIMARY KEY, "
+                           << "createdby    VARCHAR(56) " << coll
```
so we end up with a lot less code

sisuresh replied:
This makes sense to me, but I noticed that this wasn't done for the other ledger entries. I want to make sure there wasn't a specific reason for listing the fields individually before I change this. @jonjove?

jonjove replied:
The reason other ledger entries aren't like this is historical. It's possible we will move this direction in the long run. We already moved slightly in this direction with #2593.

sisuresh replied:
Makes sense. Thanks

sisuresh replied:
I made the change to use a `claimablebalanceentry` column. I also updated `db-schema.md`.

MonsieurNicolas commented on src/transactions/test/ClaimableBalanceTests.cpp:
```diff
@@ -0,0 +1,1029 @@
+
+                for (uint32_t i = 0; i < 4; i++)
+                {
+                    SECTION("predicate at level " + std::to_string(i))
```
from recent findings, let's be defensive and avoid using `to_string` -> use `fmt::format` instead

sisuresh replied:
done

sisuresh commented on src/invariant/LedgerEntryIsValid.cpp:
```diff
@@ -102,6 +102,8 @@ LedgerEntryIsValid::checkIsValid(LedgerEntry const& le, uint32_t ledgerSeq,
     case DATA:
         return checkIsValid(le.data.data(), version);
+    case CLAIMABLE_BALANCE:
+        return checkIsValid(le.data.claimableBalance(), version);
```
Added an issue - https://github.com/stellar/stellar-core/issues/2620

sisuresh commented on src/database/Database.cpp:
```diff
@@ -440,6 +440,7 @@ Database::initialize()
         mApp.getLedgerTxnRoot().dropOffers();
         mApp.getLedgerTxnRoot().dropTrustLines();
         mApp.getLedgerTxnRoot().dropData();
+        mApp.getLedgerTxnRoot().dropClaimableBalances();
```
@MonsieurNicolas I just rebased with master to pull in #2593. Is there anything else I need to do to deal with the schema upgrade for cap-0023?

sisuresh replied:
Done

MonsieurNicolas commented on src/ledger/LedgerTxnClaimableBalanceSQL.cpp:
```diff
@@ -0,0 +1,399 @@
+    mDatabase.getSession() << "CREATE TABLE claimablebalance ("
+                           << "balanceid             VARCHAR(56) " << coll
+                           << " PRIMARY KEY, "
+                           << "claimablebalanceentry TEXT " << coll
```
we should store not just `ClaimableBalanceEntry` but the full `LedgerEntry` in this column, this duplicates a small amount of data but avoids adding an extra column for `ledgerext` (this was added in the latest round of changes to the database)

sisuresh replied:
This would also allow us to remove the `lastmodified` column right?

sisuresh replied:
Nevermind. The `lastmodified` column is required.

sisuresh replied:
I made the change to store the full `LedgerEntry`.

jonjove commented on src/transactions/test/ClaimableBalanceTests.cpp:
```diff
@@ -0,0 +1,1030 @@
+    auto root = TestAccount::createRoot(*app);
+    auto const& lm = app->getLedgerManager();
+
+    const int64_t trustLineLimit = INT64_MAX;
```
Nitpick: `int64_t const`

sisuresh replied:
fixed

jonjove commented on src/transactions/test/ClaimableBalanceTests.cpp:
```diff
@@ -0,0 +1,1030 @@
+            // authorize trustline
+            if (!isNative)
+            {
+                SECTION("no trust")
```
Does this all actually work as you intend? The sequential sections with intermittent code is a pretty confusing pattern. It leaves me wondering, what _is supposed_ to happen here and what _actually does_ happen here?

Are these early sections like "no trust" and "not authorized" even required? The code in those sections should have no effect other than paying fees and consuming sequence numbers, so the sections might not really be doing anything.

In summary, I'm confused by this.

sisuresh replied:
I cleaned up the test cases and rebased into 656f3c9718fa0382f3b983b810179d689ae8e874.

sisuresh replied:
I cleaned up the tests and squashed with 572598aa0be8290908319e5d1d98e83c09c5520f.

sisuresh commented on src/ledger/test/LedgerTestUtils.cpp:
```diff
@@ -186,6 +190,21 @@ makeValid(DataEntry& d)
+    c.asset.type(ASSET_TYPE_CREDIT_ALPHANUM4);
+    strToAssetCode(c.asset.alphaNum4().assetCode, "CAD");
+
+    c.createdBy = SecretKey::random().getPublicKey();
```
The test works if I remove it. I included this so the `createdBy` key would be valid, but I guess this doesn't make a difference because we don't check for account validity in the claimable balance entry.

sisuresh commented on src/transactions/CreateClaimableBalanceOpFrame.cpp:
```diff
@@ -0,0 +1,275 @@
+int64_t
+relativeToAbsolute(TimePoint closeTime, int64_t relative)
+{
+    return closeTime > (INT64_MAX - static_cast<uint64_t>(relative))
```
Ah this is a mistake. It should be `closeTime >  static_cast<uint64_t>(INT64_MAX - relative) ? INT64_MAX
               : closeTime + relative`.

sisuresh replied:
fixed

sisuresh commented on src/transactions/test/ClaimableBalanceTests.cpp:
```diff
@@ -0,0 +1,1030 @@
+using namespace stellar::txtest;
+
+Claimant
+makeClaimant(AccountID const& account, ClaimPredicate const& pred)
```
done

sisuresh commented on src/transactions/test/ClaimableBalanceTests.cpp:
```diff
@@ -0,0 +1,1030 @@
+        break;
+    case ClaimPredicateType::CLAIM_PREDICATE_AND:
+    {
+        auto& andPredicates = pred.andPredicates();
```
Good idea. Done

sisuresh commented on src/ledger/LedgerTxnClaimableBalanceSQL.cpp:
```diff
@@ -0,0 +1,399 @@
+    mDatabase.getSession() << "CREATE TABLE claimablebalance ("
+                           << "balanceid             VARCHAR(56) " << coll
+                           << " PRIMARY KEY, "
+                           << "claimablebalanceentry TEXT " << coll
```
Looks like it's only used on the fields we use to query. Removed the collation.

sisuresh commented on src/ledger/LedgerTxnClaimableBalanceSQL.cpp:
```diff
@@ -0,0 +1,399 @@
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "crypto/KeyUtils.h"
```
removed

sisuresh commented on docs/db-schema.md:
```diff
@@ -113,6 +113,18 @@ extension | TEXT | Extension specific to DataEntry (XDR)
+
+Field | Type | Description
+------|------|---------------
+balanceid | VARCHAR(56) PRIMARY KEY | (BASE64)
```
done

sisuresh commented:
@jonjove You make a good point about the test cases. They're confusing to look at, and the pattern where I use sequential sections with intermittent code can be hard to follow. I did this to avoid duplicating code, but the end result isn't great. I'll fix this.

> I wasn't able to pick out a test that created a claimable balance entry in a transaction where the operation and transaction have different source accounts

`SECTION("validate Tx account is used in hash")` tests this, but that's obviously not clear from the name. Being able to identify any tested functionality based on the `SECTION` name is a good requirement. I'll make a pass through the test cases and try to change - 
 1. Ambiguous section names.
 2. Any Large `SECTION` that tests multiple code paths without making it clear.

jonjove commented on src/transactions/test/ClaimableBalanceTests.cpp:
```diff
@@ -0,0 +1,1040 @@
+                }
+            }
+
+            SECTION("multiple create and claims")
```
Not sure I understand the purpose of this test? Unless I'm mistaken, it effectively does "create -> claim -> fund -> create -> claim" with each step in its own transaction. Was the intention to do that as a single transaction?

sisuresh replied:
The test here was to make sure that the claim doesn't have any unintended side effects that could get in the way of another create/claim later on.

sisuresh commented on src/transactions/test/ClaimableBalanceTests.cpp:
```diff
@@ -0,0 +1,1040 @@
+                    ex_CREATE_CLAIMABLE_BALANCE_NO_TRUST);
+            }
+
+            // create and modify trustlines regardless of asset type so the
```
Ah the comment should be updated to say the number of subentries is the same in both scenarios.

sisuresh commented on src/ledger/LedgerTxnClaimableBalanceSQL.cpp:
```diff
@@ -0,0 +1,386 @@
+    mDatabase.getSession() << "CREATE TABLE claimablebalance ("
+                           << "balanceid             VARCHAR(48) " << coll
+                           << " PRIMARY KEY, "
+                           << "claimablebalanceentry TEXT NOT NULL, "
```
Changed to `ledgerentry`.

sisuresh commented:
@jonjove I made the last round of PR changes and squashed/rebased. I also added a test for creating and claiming in the same tx under the section `create and claim in same tx`.

jonjove commented:
r+ 29c1cab

MonsieurNicolas commented:
@latobarita: retry
